### DISCUSSION TITLE
"stabilityDays" is not working for Maven / Gradle Lite and Gitlab Package Registry

### DISCUSSION THREAD

**Discussion by mrclrchtr** [renovatebot/renovate#9784] [Request Help]
Hi, unfortunately it is not working that our dependencies that we host in a gitlab package registry are held back for a certain time.

I examined the debug log and could see a difference between a working dependency and ours. I think ours is missing the "releaseTimestamp". This could be the reason or? Why is that? Can I fix this? 

Ours:
```
{
                 "depName": "xxx",
                 "currentValue": "1.2.3",
                 "managerData": {
                   "fileReplacePosition": 2016,
                   "packageFile": "build.gradle.kts"
                 },
                 "fileReplacePosition": 2016,
                 "registryUrls": [
                   "https://repo.maven.apache.org/maven2",
                   "https://jcenter.bintray.com/",
                   "https://gitlab.com/api/v4/groups/xxx/-/packages/maven"
                 ],
                 "depIndex": 5,
                 "updates": [
                   {
                     "newVersion": "1.2.4",
                     "newValue": "1.2.4",
                     "bucket": "patch",
                     "newMajor": 1,
                     "newMinor": 2,
                     "updateType": "patch",
                     "branchName": "renovate/patch-xxx"
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://gitlab.com/yyy/zzz/xxx",
                 "homepage": "https://gitlab.com/yyy/zzz/xxx",
                 "currentVersion": "1.2.3",
                 "isSingleVersion": true,
                 "fixedVersion": "1.2.3"
               },
```

Working:
```
{
                 "depName": "software.amazon.awssdk:bom",
                 "currentValue": "2.16.39",
                 "managerData": {
                   "fileReplacePosition": 2128,
                   "packageFile": "build.gradle.kts"
                 },
                 "groupName": "awsSdk",
                 "fileReplacePosition": 2128,
                 "registryUrls": [
                   "https://repo.maven.apache.org/maven2",
                   "https://jcenter.bintray.com/",
                   "https://gitlab.com/api/v4/groups/xxx/-/packages/maven"
                 ],
                 "depIndex": 6,
                 "updates": [
                   {
                     "newVersion": "2.16.50",
                     "newValue": "2.16.50",
                     "bucket": "patch",
                     "newMajor": 2,
                     "newMinor": 16,
                     "updateType": "patch",
                     "releaseTimestamp": "2021-04-28T19:06:10.000Z",
                     "branchName": "renovate/patch-awssdk"
                   }
                 ],
                 "warnings": [],
                 "homepage": "https://aws.amazon.com/sdkforjava",
                 "currentVersion": "2.16.39",
                 "isSingleVersion": true,
                 "fixedVersion": "2.16.39"
               },
```

**Comment by viceice** [comment]
can you post the `npm view --json xxx time` dump? This should print all versions with there release times

https://github.com/renovatebot/renovate/blob/38620be343cca6814eb5996bb1c9b7da5a618db5/lib/datasource/npm/get.ts#L122-L124

**Reply by viceice** [reply]
I think gitlab is maybe missing those, i've seen the same on proget some time ago.

**Comment by rarkins** [comment]
If I recall correctly, Maven datasources don't have a standard for indicating timestamps so we only do it for Maven Central which we have found to return a reliable `last-modified` header: https://github.com/renovatebot/renovate/blob/a9c9d56779a3c10b3d8a26bd4eb412c674d4215b/lib/datasource/maven/util.ts#L119-L122

**Reply by viceice** [reply]
ok, maybe we can allow `last-modified` header for all if available?

(had not seen that this is maven)

**Reply by mrclrchtr** [reply]
There is definitely a `last-modified` header when I download the pom.xml from the gitlab repo. This also matches exactly the release date of the package. So this could solve the problem for GitLab.

**Reply by viceice** [reply]
So i think we can return `last-modified` for all maven repos if available. I don't think it will harm, as users can't use `stabilityDays` anyways for them.

**Reply by rarkins** [reply]
@zharinov did you find anything about these headers which could cause chaos?

**Reply by zharinov** [reply]
For Maven central, it worked surprisingly well, though we weren't sure about other sources. Hopefully, Gitlab will return good ones too.

**Reply by zharinov** [reply]
I guess, to check it out we should prepare private packages, shouldn't we? Or there are some public packages available?

**Reply by mrclrchtr** [reply]
Maybe you can use this one? https://gitlab.com/gitlab-examples/maven/simple-maven-dep/-/packages/33641

**Reply by rarkins** [reply]
We need to ask:
- What's the worst thing that can go wrong if the last-modified headers are unreliable?
- In such cases, do we have sufficient configuration options available to work around them?

I'm guessing the two major problems would be:
- No header present (which is fine.. no release time stamp)
- Header is always recent (in which case the package always seems new)

I think we're only really using it for stabilityDays now, or have I forgotten something?

**Reply by viceice** [reply]
yes, I'm pretty sure we only use it for `stabilityDays` too.

the behavior change would be with `stabilityDays`
- now: immediately passed
- with rolling header: never passed

So you can't really use `stabilityDays` either

**Reply by rarkins** [reply]
So the warning would need to be: you can't use `stabilityDays` if a maven source returns unreliable last-modified headers. We might even be able to test for that in the datasource one day too (e.g. every version returns the same or similar value) and then strip it.

**Reply by viceice** [reply]
yes, maybe we can add a warnig to the `stabilityDays` docs?

**Reply by rarkins** [reply]
Yes, we should have a warning there anyway saying that some datasources don't provide a release timestamp or we don't support it yet. And then a special note about Maven.

**Reply by viceice** [reply]
@HonkingGoose Can you propose a pr 🙃 

https://github.com/renovatebot/renovate/blob/HEAD/docs/usage/configuration-options.md#stabilitydays

**Reply by zharinov** [reply]
For GitLab own maven package example, `last-modified` header is correct

**Reply by zharinov** [reply]
This PR seems to be pretty safe to merge: https://github.com/renovatebot/renovate/pull/9791. Isn't it?

**Reply by rarkins** [reply]
I don't have an objection to supporting all maven hosts though, once the docs are updated

**Reply by HonkingGoose** [reply]
> @HonkingGoose Can you propose a pr 🙃
> 
> https://github.com/renovatebot/renovate/blob/HEAD/docs/usage/configuration-options.md#stabilitydays

I made PR #9792 to improve the docs (which is merged now).

**Reply by rarkins** [reply]
Docs are updated, @zharinov let's use last-modified for all maven sources now

**Reply by zharinov** [reply]
#9795

**Reply by mrclrchtr** [reply]
Thank you very much for the quick reply and the handling of the problem!
